### PR TITLE
AOD: fixe memory leak + allow overwriting metadata

### PR DESCRIPTION
**Pull request by shahor02**

### CODE REVIEW

File: Framework/Core/src/CommonDataProcessors.cxx [hunk 1]
```diff
@@ -354,7 +354,7 @@ DataProcessorSpec
         auto s = pc.inputs().get<TableConsumer>(ref.spec->binding);
         auto table = s->asArrowTable();
         if (!table->Validate().ok()) {
-          LOGP(WARNING, "The table \"{}\" is not valid and will not be saved!", tableName);
+          LOGP(warnING, "The table \"{}\" is not valid and will not be saved!", tableName);
```
Comment (sawenzel) [inline Framework/Core/src/CommonDataProcessors.cxx]: is that a typo? `LOGP(warn)` ?

Comment (ktf) [inline Framework/Core/src/CommonDataProcessors.cxx]:
```suggestion
          LOGP(warning, "The table \"{}\" is not valid and will not be saved!", tableName);
```

File: Framework/Core/src/CommonDataProcessors.cxx [hunk 2]
```diff
@@ -383,7 +383,7 @@ DataProcessorSpec
           } else {
             ta2tr.addAllBranches();
           }
-          ta2tr.process();
+          delete ta2tr.process();
```
Comment (sawenzel) [inline Framework/Core/src/CommonDataProcessors.cxx]: What is the problem here? `ta2tr` does not seem to be a pointer. Is this deleting an object returned by `process()` which is not **owned** by `ta2tr`? This looks a bit awkward.

**Comment by jgrosseo** [comment]
Hi Ruben,
Why we need the case of overwriting the metadata?

File: Framework/Core/src/CommonDataProcessors.cxx (hunk 1 above)
Comment (shahor02) [reply Framework/Core/src/CommonDataProcessors.cxx]: It is actually a feature of the upper to lower case converter: it checks only for `|LOGP[(]WARN|LOGP(warn|g;`
@davidrohr can this be fixed?

**Comment by shahor02** [comment]
@jgrosseo because currently, it produces FATAL when metadata is there

File: Framework/Core/src/CommonDataProcessors.cxx (hunk 2 above)
Comment (shahor02) [reply Framework/Core/src/CommonDataProcessors.cxx]: TableToTree returns the pointer to the tree after the writing. It has only a default constructor, so although it has a pointer data member, it is not deleted. The alternative would be to define it as unique_ptr, but then its potential consumer should be aware of this.

File: Framework/Core/src/CommonDataProcessors.cxx (hunk 1 above)
Comment (shahor02) [reply Framework/Core/src/CommonDataProcessors.cxx]: @ktf I've learned that this case converter is yours, can you add `WARNING`?

**Comment by jgrosseo** [comment]
> @jgrosseo because currently, it produces FATAL when metadata is there

Yes, that is not good of course. I would simply change that to WARNING then. The overwrite does not harm, but does not help either...

**Comment by shahor02** [comment]
@jgrosseo do you want simply skip creating the metada if it is already there?

**Comment by shahor02** [comment]
@sawenzel @ktf I've changed the raw mTree pointer in the TableToTree to shared one and removed the deletion of the returned pointer.

**Comment by ktf** [comment]
@shahor02 i would actually check if we cannot simply avoid returning the tree.

**Comment by jgrosseo** [comment]
> @jgrosseo do you want simply skip creating the metada if it is already there?

Yes, exactly. It will be identical (unless we have a quite bad setup...)

**Comment by shahor02** [comment]
@ktf check with whom? At the moment the returned pointer is never used, but I understood the returning was done on purpose in case somebody will need this pointer.

**Comment by shahor02** [comment]
@jgrosseo ok, I've changed the code to preserve the existing one

**Comment by ktf** [comment]
@shahor02 what i meant is that the pointer might have been used, so before changing it to anything else, we should have checked. If it's not used, I would be in favour of simply deleting the object inside the function and not return anything.

**Comment by shahor02** [comment]
Hi @ktf, currently the pointer is not used in any real code, but is used in the test: https://github.com/AliceO2Group/AliceO2/blob/b82fa607e5278a265da5d04c9c25231873795b56/Framework/Core/test/test_TreeToTable.cxx#L150-L154
Let me know if you want to preserve it.
In any case, to have this fix in the nightly, I would merge this PR once the CIs are green, then, if decided, can avoid returning a pointer and implement a destructor.

**Comment by ktf** [comment]
yeah... I guess the correct thing is to return a unique ptr, then. We can merge this and think about it later.

**Comment by shahor02** [comment]
> yeah... I guess the correct thing is to return a unique ptr, then. We can merge this and think about it later.

For the record: at the moment it returns shared ptr, which is destroyed since never assigned.

**Comment by ktf** [comment]
Yes, sure, that would work as well. I guess the unique_ptr is maybe better / safer, but we can always change it later.

**Pull request merged**